A6.md

```diff
@@ -0,0 +1,451 @@
+
+![State Diagram](A6_graphics/StateDiagram.png)
+
+### Retry Internals
+
+#### Where Retries Occur
+
+The retry policy will be implemented in-between the channel and the load balancing policy. That way every retry gets a chance to be sent out on a different subchannel than it originally failed on.
+
+![Where Retries Occur](A6_graphics/WhereRetriesOccur.png)
+
+#### When Retries are Valid
+
+In certain cases it is not valid to retry an RPC. These cases occur when the RPC has been *committed*, and thus it does not make sense to perform the retry.
+
+An RPC becomes *committed* in two scenarios:
+
+1. The client receives a non-error response (either an explicit OK status or any response message) from the server.
+2. The client’s outgoing message has overflowed the gRPC client library’s buffer.
+
+The reasoning behind the first scenario is that as soon as a non-error response is seen from the server, it is transmitted up to the client application. This may fundamentally change the state of the client, so we cannot safely retry if a failure occurs later in the RPC’s life.
+
+To clarify the second scenario, we define an *outgoing message* as everything the client sends on its connection to the server. For unary and server streaming calls, the outgoing message is a single message. For client and bidirectional streaming calls, the outgoing message is the entire message stream issued by the client after opening the connection. The gRPC client library buffers outgoing messages, and as long as the entirety of the outgoing message is in the buffer, it can be resent and retried. But as soon as the outgoing message grows too large to buffer, the gRPC client library cannot replay the entire stream of messages, and thus retries are not valid.
+
+#### Buffered RPCs
+
+The gRPC library will have a configurable amount of available memory, `retry_buffer_size_in_bytes`, to buffer outgoing retryable or hedged RPCs. There is also a per-RPC size limit ,`per_rpc_buffer_limit_in_bytes`. These limits are configured by the client, rather than coming from the service config.
+
+RPCs may only be retried when they are contained in the buffer. 
+
+Hedged RPCs with a non-zero delay will only send out the subsequent hedged requests when the original RPC remains in the buffer. For hedge policies with a delay of zero milliseconds, the RPC does not need to be placed in the buffer at all and the outgoing RPCs can be sent immediately then discarded. 
+
+After the RPC response has been returned to the client application layer, the RPC is removed from the buffer.
+
+New RPCs which do not fit in the available buffer space (either due to the total available buffer space, or due to the per-RPC limit) will not be retryable. For client-side streaming RPCs, if additional outgoing messages would cause the buffered size to exceed the per-RPC limit, the entire message will be removed from the buffer. 
```

Comment:
> New RPCs which do not fit in the available buffer space (either due to the total available buffer space, or due to the per-RPC limit) will not be retryable.

Does that mean that they'll fail immediately? Does that mean they will make original request and in case of failure won't qualify for retry and fail after first initial request? I assume it's the former, maybe worth clarifying?